### PR TITLE
Allow users to specify deletion propagation for prune

### PR DESCRIPTION
**Pull request by mortent**
Allow users to specify the deletion propagation policy for pruning.

@seans3 @phanimarupaka @monopole

### CODE REVIEW

File: pkg/apply/prune/prune.go [hunk 1]
```diff
@@ -190,13 +190,16 @@ type Options struct {
...
+
+	PropagationPolicy metav1.DeletionPropagation
 }
 
 // Prune deletes the set of resources which were previously applied
 // (retrieved from previous inventory objects) but omitted in
 // the current apply. Prune also delete all previous inventory
 // objects. Returns an error if there was a problem.
 func (po *PruneOptions) Prune(currentObjects []*resource.Info, eventChannel chan<- event.Event, o Options) error {
+	fmt.Printf("Pruning with prop policy %s\n", string(o.PropagationPolicy))
```
Comment (seans3) [inline pkg/apply/prune/prune.go:202]: Should this be removed?

File: cmd/apply/cmdapply.go [hunk 2]
```diff
@@ -68,14 +72,20 @@ type ApplyRunner struct {
...
+	wait                   bool
+	period                 time.Duration
+	timeout                time.Duration
+	noPrune                bool
+	prunePropagationPolicy string
 }
 
-func (r *ApplyRunner) Run(cmd *cobra.Command, args []string) {
+func (r *ApplyRunner) RunE(cmd *cobra.Command, args []string) error {
+	prunePropPolicy, err := convertPropagationPolicy(r.prunePropagationPolicy)
```
Comment (seans3) [inline cmd/apply/cmdapply.go:84]: Should we add a comment somewhere that the default propagation policy is `background`

File: pkg/apply/task/prune_task.go [hunk 3]
```diff
@@ -13,9 +14,10 @@ import (
 // by using the PruneOptions. The provided Objects is the
 // set of resources that have just been applied.
 type PruneTask struct {
-	PruneOptions *prune.PruneOptions
-	Objects      []*resource.Info
-	DryRun       bool
+	PruneOptions      *prune.PruneOptions
+	Objects           []*resource.Info
+	DryRun            bool
+	PropagationPolicy metav1.DeletionPropagation
```
Comment (seans3) [inline pkg/apply/task/prune_task.go:20]: Should this be separate from the `PruneOptions` or part of it?

**Comment by seans3** [comment]
Looks good. Thanks for taking the initiative on this. I had a couple questions.

File: pkg/apply/prune/prune.go (hunk 1 above)
Comment (mortent) [reply pkg/apply/prune/prune.go]: Yeah, this shouldn't be here..

File: cmd/apply/cmdapply.go (hunk 2 above)
Comment (mortent) [reply cmd/apply/cmdapply.go:84]: We are setting the default value in the cobra command. So running `apply --help` will show that the default is `Background`. I have also updated this PR so the Options passed into the Applier will also handle defaulting to Background correctly.

File: pkg/apply/task/prune_task.go (hunk 3 above)
Comment (mortent) [reply pkg/apply/task/prune_task.go:20]: So we are passing the PropagationPolicy into the `Prune` function as part of the options parameter object. I prefer this solution over setting values directly on the `PruneOptions`. After we moved `DryRun` from being a field on `PruneOptions` to being passed into the `Prune` function, the Applier only interacts with the `PruneOptions` through functions. This means we can use an interface to help with testing.

**Pull request merged**